File: app/controllers/ExaminationEventController.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import controllers.base.BaseController;
import io.ebean.DB;
import io.ebean.ExpressionList;
import io.ebean.text.PathProperties;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;
import javax.inject.Inject;
import models.Exam;
import models.ExaminationDate;
import models.ExaminationEvent;
import models.ExaminationEventConfiguration;
import models.calendar.MaintenancePeriod;
import org.joda.time.DateTime;
import org.joda.time.Interval;
import org.joda.time.LocalDate;
import org.joda.time.format.ISODateTimeFormat;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.With;
import sanitizers.Attrs;
import sanitizers.ExaminationDateSanitizer;
import sanitizers.ExaminationEventSanitizer;
import util.config.ByodConfigHandler;
import util.config.ConfigReader;

public class ExaminationEventController extends BaseController {

    private final Logger logger = LoggerFactory.getLogger(ExaminationEventController.class);

    @Inject
    ByodConfigHandler byodConfigHandler;

    @Inject
    ConfigReader configReader;

    // PRINTOUT EXAM RELATED -->
    @With(ExaminationDateSanitizer.class)
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result insertExaminationDate(Long eid, Http.Request request) {
        Exam exam = DB.find(Exam.class, eid);
        if (exam == null) {
            return notFound("exam not found");
        }
        LocalDate date = request.attrs().get(Attrs.DATE);
        ExaminationDate ed = new ExaminationDate();
        ed.setDate(date.toDate());
        ed.setExam(exam);
        ed.save();
        return ok(ed);
    }

    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result removeExaminationDate(Long id, Long edid) {
        ExaminationDate ed = DB.find(ExaminationDate.class, edid);
        if (ed == null) {
            return notFound("examination date not found");
        }
        ed.delete();
        return ok();
    }

    // <--
    private DateTime getEventEnding(ExaminationEvent ee) {
        ExaminationEventConfiguration config = ee.getExaminationEventConfiguration();
        if (config == null) {
            return ee.getStart();
        }
        return ee.getStart().plusMinutes(config.getExam().getDuration());
    }

    private int getParticipantUpperBound(DateTime start, DateTime end, Long id) {
        ExpressionList<ExaminationEvent> el = DB.find(ExaminationEvent.class).where().le("start", end);
        if (id != null) {
            el = el.ne("id", id);
        }
        return el
            .findSet()
            .stream()
            .filter(ee -> !getEventEnding(ee).isBefore(start))
            .mapToInt(ExaminationEvent::getCapacity)
            .sum();
    }

    private boolean isWithinMaintenancePeriod(Interval interval) {
        return DB
            .find(MaintenancePeriod.class)
            .findSet()
            .stream()
            .map(p -> new Interval(p.getStartsAt(), p.getEndsAt()))
            .anyMatch(i -> i.overlaps(interval));
    }

    @With(ExaminationEventSanitizer.class)
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result insertExaminationEvent(Long eid, Http.Request request) {
        Exam exam = DB.find(Exam.class, eid);
        if (exam == null) {
            return notFound("exam not found");
        }
        ExaminationEventConfiguration eec = new ExaminationEventConfiguration();
        // HOX! maybe in the future this can be some preset event shared by multiple exams.
        // For time being lets always create new events.
        ExaminationEvent ee = new ExaminationEvent();
        DateTime start = request.attrs().get(Attrs.START_DATE);
        if (start.isBeforeNow()) {
            return forbidden("i18n_error_examination_event_in_the_past");
        }
        DateTime end = start.plusMinutes(exam.getDuration());
        if (isWithinMaintenancePeriod(new Interval(start, end))) {
            return forbidden("i18n_error_conflicts_with_maintenance_period");
        }
        int ub = getParticipantUpperBound(start, end, null);
        int capacity = request.attrs().get(Attrs.CAPACITY);
        if (capacity + ub > configReader.getMaxByodExaminationParticipantCount()) {
            return forbidden("i18n_error_max_capacity_exceeded");
        }
        String quitPassword = request.attrs().get(Attrs.QUIT_PASSWORD);
        if (exam.getImplementation() == Exam.Implementation.CLIENT_AUTH && quitPassword == null) {
            return forbidden("no quit password provided");
        }
        String settingsPassword = request.attrs().get(Attrs.SETTINGS_PASSWORD);
        if (exam.getImplementation() == Exam.Implementation.CLIENT_AUTH && settingsPassword == null) {
            return forbidden("no settings password provided");
        }
        ee.setStart(start);
        ee.setDescription(request.attrs().get(Attrs.DESCRIPTION));
        ee.setCapacity(request.attrs().get(Attrs.CAPACITY));
        ee.save();
        eec.setExaminationEvent(ee);
        eec.setExam(exam);
        eec.setHash(UUID.randomUUID().toString());
        if (quitPassword != null && settingsPassword != null) {
            encryptQuitPassword(eec, quitPassword);
            encryptSettingsPassword(eec, settingsPassword, quitPassword);
            // Pass back the plaintext password, so it can be shown to user
            eec.setQuitPassword(quitPassword);
            eec.setSettingsPassword(settingsPassword);
        }
        eec.save();
        return ok(eec);
    }

    @With(ExaminationEventSanitizer.class)
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result updateExaminationEvent(Long eid, Long eecid, Http.Request request) {
        Exam exam = DB.find(Exam.class, eid);
        Optional<ExaminationEventConfiguration> oeec = DB
            .find(ExaminationEventConfiguration.class)
            .where()
            .idEq(eecid)
            .eq("exam.id", eid)
            .findOneOrEmpty();
        if (exam == null || oeec.isEmpty()) {
            return notFound("event not found");
        }
        ExaminationEventConfiguration eec = oeec.get();
        boolean hasEnrolments = !eec.getExamEnrolments().isEmpty();
        ExaminationEvent ee = eec.getExaminationEvent();
        String quitPassword = request.attrs().get(Attrs.QUIT_PASSWORD);
        if (eec.getExam().getImplementation() == Exam.Implementation.CLIENT_AUTH && quitPassword == null) {
            return forbidden("no quit password provided");
        }
        String settingsPassword = request.attrs().get(Attrs.SETTINGS_PASSWORD);
        if (eec.getExam().getImplementation() == Exam.Implementation.CLIENT_AUTH && settingsPassword == null) {
            return forbidden("no settings password provided");
        }
        DateTime start = request.attrs().get(Attrs.START_DATE);
        if (!hasEnrolments) {
            if (start.isBeforeNow()) {
                return forbidden("i18n_error_examination_event_in_the_past");
            }
            ee.setStart(start);
        }
        DateTime end = start.plusMinutes(exam.getDuration());
        if (isWithinMaintenancePeriod(new Interval(start, end))) {
            return forbidden("i18n_error_conflicts_with_maintenance_period");
        }
        int ub = getParticipantUpperBound(start, end, ee.getId());
        int capacity = request.attrs().get(Attrs.CAPACITY);
        if (capacity + ub > configReader.getMaxByodExaminationParticipantCount()) {
            return forbidden("i18n_error_max_capacity_exceeded");
        }
        ee.setCapacity(capacity);
        ee.setDescription(request.attrs().get(Attrs.DESCRIPTION));
        ee.update();
        if (quitPassword == null || settingsPassword == null) {
            return ok(eec);
        }
        if (!hasEnrolments) {
            encryptQuitPassword(eec, quitPassword);
            encryptSettingsPassword(eec, settingsPassword, quitPassword);
            eec.save();
            // Pass back the plaintext passwords, so they can be shown to user
            eec.setQuitPassword(quitPassword);
            eec.setSettingsPassword(settingsPassword);
        } else {
            // Disallow changing password if enrolments exist for this event
            // Pass back the original unchanged passwords
            eec.setQuitPassword(
                byodConfigHandler.getPlaintextPassword(eec.getEncryptedQuitPassword(), eec.getQuitPasswordSalt())
            );
            eec.setSettingsPassword(
                byodConfigHandler.getPlaintextPassword(
                    eec.getEncryptedSettingsPassword(),
                    eec.getSettingsPasswordSalt()
                )
            );
        }
        return ok(eec);
    }

    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result removeExaminationEvent(Long eid, Long eeid) {
        Optional<ExaminationEventConfiguration> oeec = DB
            .find(ExaminationEventConfiguration.class)
            .where()
            .idEq(eeid)
            .eq("exam.id", eid)
            .findOneOrEmpty();
        Exam exam = DB.find(Exam.class, eid);
        if (oeec.isEmpty() || exam == null) {
            return notFound("event not found");
        }
        ExaminationEventConfiguration eec = oeec.get();
        if (!eec.getExamEnrolments().isEmpty()) {
            return forbidden("event can not be deleted because there are enrolments involved");
        }
        eec.delete();
        // Check if we can delete the event altogether (in case no configs are using it)
        Set<ExaminationEventConfiguration> configs = DB
            .find(ExaminationEventConfiguration.class)
            .where()
            .eq("examinationEvent", eec.getExaminationEvent())
            .findSet();
        if (configs.isEmpty()) {
            eec.getExaminationEvent().delete();
        }
        return ok();
    }

    private void encryptSettingsPassword(ExaminationEventConfiguration eec, String password, String quitPassword) {
        try {
            String oldPwd = eec.getEncryptedSettingsPassword() != null
                ? byodConfigHandler.getPlaintextPassword(
                    eec.getEncryptedSettingsPassword(),
                    eec.getSettingsPasswordSalt()
                )
                : null;

            if (!password.equals(oldPwd)) {
                String newSalt = UUID.randomUUID().toString();
                eec.setEncryptedSettingsPassword(byodConfigHandler.getEncryptedPassword(password, newSalt));
                eec.setSettingsPasswordSalt(newSalt);
                // Pre-calculate config key, so we don't need to do it each time a check is needed
                eec.setConfigKey(byodConfigHandler.calculateConfigKey(eec.getHash(), quitPassword));
            }
        } catch (Exception e) {
            logger.error("unable to set settings password", e);
            throw new RuntimeException(e);
        }
    }

    private void encryptQuitPassword(ExaminationEventConfiguration eec, String password) {
        try {
            String oldPwd = eec.getEncryptedQuitPassword() != null
                ? byodConfigHandler.getPlaintextPassword(eec.getEncryptedQuitPassword(), eec.getQuitPasswordSalt())
                : null;

            if (!password.equals(oldPwd)) {
                String newSalt = UUID.randomUUID().toString();
                eec.setEncryptedQuitPassword(byodConfigHandler.getEncryptedPassword(password, newSalt));
                eec.setQuitPasswordSalt(newSalt);
                // Pre-calculate config key, so we don't need to do it each time a check is needed
                eec.setConfigKey(byodConfigHandler.calculateConfigKey(eec.getHash(), password));
            }
        } catch (Exception e) {
            logger.error("unable to set settings password", e);
            throw new RuntimeException(e);
        }
    }

    @Restrict({ @Group("ADMIN") })
    public Result listExaminationEvents(Optional<String> start, Optional<String> end) {
        PathProperties pp = PathProperties.parse(
            "(*, exam(*, course(*), examOwners(*)), examinationEvent(*), examEnrolments(*))"
        );
        ExpressionList<ExaminationEventConfiguration> query = DB
            .find(ExaminationEventConfiguration.class)
            .apply(pp)
            .where();
        if (start.isPresent()) {
            DateTime startDate = DateTime.parse(start.get(), ISODateTimeFormat.dateTimeParser());
            query = query.ge("examinationEvent.start", startDate.toDate());
        }
        if (end.isPresent()) {
            DateTime endDate = DateTime.parse(end.get(), ISODateTimeFormat.dateTimeParser());
            query = query.lt("examinationEvent.start", endDate.toDate());
        }
        Set<ExaminationEventConfiguration> exams = query.where().eq("exam.state", Exam.State.PUBLISHED).findSet();
        return ok(exams, pp);
    }

    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result listOverlappingExaminationEvents(String start, Integer duration) {
        PathProperties pp = PathProperties.parse("(*, examinationEventConfiguration(exam(id, duration)))");
        DateTime startDate = DateTime.parse(start, ISODateTimeFormat.dateTimeParser());
        DateTime endDate = startDate.plusMinutes(duration);
        Set<ExaminationEvent> events = DB
            .find(ExaminationEvent.class)
            .where()
            .le("start", endDate)
            .findSet()
            .stream()
            .filter(ee -> !getEventEnding(ee).isBefore(startDate))
            .collect(Collectors.toSet());
        return ok(events, pp);
    }
}

File: app/controllers/enrolment/ExaminationEventController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.enrolment

import controllers.base.scala.ExamBaseController
import io.ebean.DB
import io.ebean.text.PathProperties
import miscellaneous.config.{ByodConfigHandler, ConfigReader}
import miscellaneous.scala.DbApiHelper
import models.calendar.MaintenancePeriod
import models.enrolment.{ExaminationEvent, ExaminationEventConfiguration}
import models.exam.{Exam, ExaminationDate}
import models.user.Role
import org.joda.time.format.ISODateTimeFormat
import org.joda.time.{DateTime, Interval}
import play.api.Logging
import play.api.mvc.*
import security.scala.Auth.{AuthenticatedAction, authorized}
import validation.scala.core.{ScalaAttrs, Validators}
import validation.scala.exam.{ExaminationDateValidator, ExaminationEventValidator}

import java.util.UUID
import javax.inject.Inject

class ExaminationEventController @Inject() (
    byodConfigHandler: ByodConfigHandler,
    configReader: ConfigReader,
    authenticated: AuthenticatedAction,
    validators: Validators,
    val controllerComponents: ControllerComponents
)(implicit ec: scala.concurrent.ExecutionContext)
    extends BaseController
    with DbApiHelper
    with ExamBaseController
    with Logging:

  // PRINTOUT EXAM RELATED -->
  def insertExaminationDate(eid: Long): Action[AnyContent] =
    authenticated
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT)))
      .andThen(validators.validated(ExaminationDateValidator)) { request =>
        Option(DB.find(classOf[Exam], eid)) match
          case None => NotFound("exam not found")
          case Some(exam) =>
            val date = request.attrs(ScalaAttrs.DATE)
            val ed   = new ExaminationDate()
            ed.setDate(date.toDate)
            ed.setExam(exam)
            ed.save()
            Ok(ed.asJson)
      }

  def removeExaminationDate(id: Long, edid: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT))) { _ =>
      Option(DB.find(classOf[ExaminationDate], edid)) match
        case None => NotFound("examination date not found")
        case Some(ed) =>
          ed.delete()
          Ok
    }

  // <--
  private def getEventEnding(ee: ExaminationEvent): DateTime =
    Option(ee.getExaminationEventConfiguration) match
      case None         => ee.getStart
      case Some(config) => ee.getStart.plusMinutes(config.getExam.getDuration)

  private def getParticipantUpperBound(start: DateTime, end: DateTime, id: Option[Long]): Int =
    var query = DB.find(classOf[ExaminationEvent]).where().le("start", end)
    id.foreach(i => query = query.ne("id", i))

    query.distinct
      .filter(ee => !getEventEnding(ee).isBefore(start))
      .map(_.getCapacity)
      .sum

  private def isWithinMaintenancePeriod(interval: Interval): Boolean =
    DB.find(classOf[MaintenancePeriod])
      .distinct
      .exists { p =>
        val maintenanceInterval = new Interval(p.getStartsAt, p.getEndsAt)
        maintenanceInterval.overlaps(interval)
      }

  def insertExaminationEvent(eid: Long): Action[AnyContent] =
    authenticated
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT)))
      .andThen(validators.validated(ExaminationEventValidator)) { request =>
        Option(DB.find(classOf[Exam], eid)) match
          case None       => NotFound("exam not found")
          case Some(exam) =>
            // Note: Validation is handled by ExaminationEventValidator
            val dto   = request.attrs(ScalaAttrs.EXAMINATION_EVENT)
            val start = dto.start

            if start.isBeforeNow then Forbidden("i18n_error_examination_event_in_the_past")
            else
              val end = start.plusMinutes(exam.getDuration)
              if isWithinMaintenancePeriod(new Interval(start, end)) then
                Forbidden("i18n_error_conflicts_with_maintenance_period")
              else
                val ub       = getParticipantUpperBound(start, end, None)
                val capacity = dto.capacity
                if capacity + ub > configReader.getMaxByodExaminationParticipantCount then
                  Forbidden("i18n_error_max_capacity_exceeded")
                else
                  val quitPassword = dto.quitPassword
                  if exam.getImplementation == Exam.Implementation.CLIENT_AUTH && quitPassword.isEmpty then
                    Forbidden("no quit password provided")
                  else
                    val settingsPassword = dto.settingsPassword
                    if exam.getImplementation == Exam.Implementation.CLIENT_AUTH && settingsPassword.isEmpty then
                      Forbidden("no settings password provided")
                    else
                      val eec = new ExaminationEventConfiguration()
                      val ee  = new ExaminationEvent()
                      ee.setStart(start)
                      ee.setDescription(dto.description)
                      ee.setCapacity(dto.capacity)
                      ee.save()
                      eec.setExaminationEvent(ee)
                      eec.setExam(exam)
                      eec.setHash(UUID.randomUUID().toString)

                      for
                        qPwd <- quitPassword
                        sPwd <- settingsPassword
                      yield
                        encryptQuitPassword(eec, qPwd)
                        encryptSettingsPassword(eec, sPwd, qPwd)
                        // Pass back the plaintext password, so it can be shown to the user
                        eec.setQuitPassword(qPwd)
                        eec.setSettingsPassword(sPwd)

                      eec.save()
                      Ok(eec.asJson)
      }

  def updateExaminationEvent(eid: Long, eecid: Long): Action[AnyContent] =
    authenticated
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT)))
      .andThen(validators.validated(ExaminationEventValidator)) { request =>
        val examOpt = Option(DB.find(classOf[Exam], eid))
        val eecOpt = DB
          .find(classOf[ExaminationEventConfiguration])
          .where()
          .idEq(eecid)
          .eq("exam.id", eid)
          .find

        (examOpt, eecOpt) match
          case (Some(exam), Some(eec)) =>
            // Note: Validation is handled by ExaminationEventValidator
            val dto           = request.attrs(ScalaAttrs.EXAMINATION_EVENT)
            val hasEnrolments = !eec.getExamEnrolments.isEmpty
            val ee            = eec.getExaminationEvent
            val quitPassword  = dto.quitPassword

            if eec.getExam.getImplementation == Exam.Implementation.CLIENT_AUTH && quitPassword.isEmpty then
              Forbidden("no quit password provided")
            else
              val settingsPassword = dto.settingsPassword
              if eec.getExam.getImplementation == Exam.Implementation.CLIENT_AUTH && settingsPassword.isEmpty then
                Forbidden("no settings password provided")
              else
                val start = dto.start
                if !hasEnrolments && start.isBeforeNow then Forbidden("i18n_error_examination_event_in_the_past")
                else
                  if !hasEnrolments then ee.setStart(start)

                  val end = start.plusMinutes(exam.getDuration)
                  if isWithinMaintenancePeriod(new Interval(start, end)) then
                    Forbidden("i18n_error_conflicts_with_maintenance_period")
                  else
                    val ub       = getParticipantUpperBound(start, end, Some(ee.getId))
                    val capacity = dto.capacity
                    if capacity + ub > configReader.getMaxByodExaminationParticipantCount then
                      Forbidden("i18n_error_max_capacity_exceeded")
                    else
                      ee.setCapacity(capacity)
                      ee.setDescription(dto.description)
                      ee.update()

                      (quitPassword, settingsPassword) match
                        case (Some(qPwd), Some(sPwd)) if !hasEnrolments =>
                          encryptQuitPassword(eec, qPwd)
                          encryptSettingsPassword(eec, sPwd, qPwd)
                          eec.save()
                          // Pass back the plaintext passwords
                          eec.setQuitPassword(qPwd)
                          eec.setSettingsPassword(sPwd)
                        case (Some(_), Some(_)) =>
                          // Disallow changing password if enrolments exist - pass back original unchanged passwords
                          eec.setQuitPassword(
                            byodConfigHandler
                              .getPlaintextPassword(eec.getEncryptedQuitPassword, eec.getQuitPasswordSalt)
                          )
                          eec.setSettingsPassword(
                            byodConfigHandler.getPlaintextPassword(
                              eec.getEncryptedSettingsPassword,
                              eec.getSettingsPasswordSalt
                            )
                          )
                        case _ => ()

                      Ok(eec.asJson)
          case _ => NotFound("event not found")
      }

  def removeExaminationEvent(eid: Long, eeid: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT))) { _ =>
      val eecOpt  = DB.find(classOf[ExaminationEventConfiguration]).where().idEq(eeid).eq("exam.id", eid).find
      val examOpt = Option(DB.find(classOf[Exam], eid))

      (eecOpt, examOpt) match
        case (Some(eec), Some(_)) =>
          if !eec.getExamEnrolments.isEmpty then
            Forbidden("event can not be deleted because there are enrolments involved")
          else
            eec.delete()
            // Check if we can delete the event altogether (in case no configs are using it)
            val configs = DB
              .find(classOf[ExaminationEventConfiguration])
              .where()
              .eq("examinationEvent", eec.getExaminationEvent)
              .distinct

            if configs.isEmpty then eec.getExaminationEvent.delete()
            Ok
        case _ => NotFound("event not found")
    }

  private def encryptSettingsPassword(
      eec: ExaminationEventConfiguration,
      password: String,
      quitPassword: String
  ): Unit =
    try
      val oldPwd = Option(eec.getEncryptedSettingsPassword).map { encrypted =>
        byodConfigHandler.getPlaintextPassword(encrypted, eec.getSettingsPasswordSalt)
      }

      if !oldPwd.contains(password) then
        val newSalt = UUID.randomUUID().toString
        eec.setEncryptedSettingsPassword(byodConfigHandler.getEncryptedPassword(password, newSalt))
        eec.setSettingsPasswordSalt(newSalt)
        // Pre-calculate the config key
        eec.setConfigKey(byodConfigHandler.calculateConfigKey(eec.getHash, quitPassword))
    catch
      case e: Exception =>
        logger.error("unable to set settings password", e)
        throw new RuntimeException(e)

  private def encryptQuitPassword(eec: ExaminationEventConfiguration, password: String): Unit =
    try
      val oldPwd = Option(eec.getEncryptedQuitPassword).map { encrypted =>
        byodConfigHandler.getPlaintextPassword(encrypted, eec.getQuitPasswordSalt)
      }

      if !oldPwd.contains(password) then
        val newSalt = UUID.randomUUID().toString
        eec.setEncryptedQuitPassword(byodConfigHandler.getEncryptedPassword(password, newSalt))
        eec.setQuitPasswordSalt(newSalt)
        // Pre-calculate the config key
        eec.setConfigKey(byodConfigHandler.calculateConfigKey(eec.getHash, password))
    catch
      case e: Exception =>
        logger.error("unable to set quit password", e)
        throw new RuntimeException(e)

  def listExaminationEvents(start: Option[String], end: Option[String]): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      val pp = PathProperties.parse(
        "(*, exam(*, course(*), examOwners(*)), examinationEvent(*), examEnrolments(*))"
      )
      var query = DB.find(classOf[ExaminationEventConfiguration]).apply(pp).where()

      start.foreach { s =>
        val startDate = DateTime.parse(s, ISODateTimeFormat.dateTimeParser())
        query = query.ge("examinationEvent.start", startDate.toDate)
      }

      end.foreach { e =>
        val endDate = DateTime.parse(e, ISODateTimeFormat.dateTimeParser())
        query = query.lt("examinationEvent.start", endDate.toDate)
      }

      val exams = query.eq("exam.state", Exam.State.PUBLISHED).distinct
      ok(exams, pp)
    }

  def listOverlappingExaminationEvents(start: String, duration: Int): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT))) { _ =>
      val startDate = DateTime.parse(start, ISODateTimeFormat.dateTimeParser())
      val endDate   = startDate.plusMinutes(duration)
      val pp        = PathProperties.parse("(*, examinationEventConfiguration(exam(id, duration)))")

      val events = DB
        .find(classOf[ExaminationEvent])
        .where()
        .le("start", endDate)
        .distinct
        .filter(ee => !getEventEnding(ee).isBefore(startDate))

      ok(events, pp)
    }
